### DISCUSSION TITLE
MAIANA WiFi adapter

### DISCUSSION THREAD

**Discussion by peterantypas** [peterantypas/maiana#118] [Ideas]
After many delays I have just started working on MAIANA's WiFi adapter. This is based on the ESP32-Pico-Mini-02 module and among other things it will present a web interface for configuration:

![IMG_1767](https://user-images.githubusercontent.com/1565933/205510546-8926dfdd-eaad-41e9-a5a9-e1373f4aafc8.PNG)

There are two main architectures I'm considering:

**Option 1: A simple WiFi adapter**

This will look similar to the existing NMEA adapters but it will host an ESP32 module. It will allow MAIANA to either provide a WiFi network for NMEA data or join an existing network and push NMEA data to it. I think this may actually simplify OpenPlotter integration, as some folks seem to have problems with configuring the USB/serial interface. Mainly, though, I think it's ideal for those who don't have a WiFi system on board and would like to integrate MAIANA with an iPad or laptop without having to pay a premium for a Raspberry Pi.

Pros:
- The adapter can be added any time after installation
- Easy to do a "hard reset" of the ESP32 if something goes wrong with network configuration

Cons:
- Still needs a Cat5 cable to wire into the cabin

**Option 2: A WiFi interface on the transponder itself**

I would either add the ESP32 module to the main MAIANA board itself, or create a small adapter board that can live inside the enclosure directly attached to the Cat5 connector. The transponder would still need 12V power but everything else would be done via WiFi. 

Pros: 
- Simpler wiring - no Cat5 cable to feed to the cabin
- Easier to install on a masthead 

Cons:
- Difficult to do a "hard reset" of the ESP32 if physical access is limited and something goes wrong with configuration

Any input welcome!

**Comment by marcelrv** [comment]
This is great news.!
Now indeed the choice is tricky.
But having several esp32 onboard, I would say that the simplicity of the wiring (only power) & easier placement and the improved reception (as esp would not be tucked somewhere deep inside the hull ) I would give preference to option 2.

**Comment by kagouraki** [comment]
Very great news Peter ! I would support both options and everyone can choose. 
The software will be the same on the esp. So if option A nothing to change in the mechanical part and if option B then the pvc tube will be 30-50mm longer. 
Actually when i think of it, it is not so difficult! I have built a wifi to nmea 183 and nmea 2000 box with esp for norbert walter's yachta anemometer project. In this one i can for sure contribute. I used the wifimanager library and the nmea2000 library. If you have some draft code i can help!

**Comment by muebau** [comment]
My wife and me started the exact same thing last Christmas but did not manage to give it the last finish this year.

https://github.com/muebau/MAIANA-WiFi

I think there might be some useful ideas in our project (get time from GPS etc.)

**Reply by kagouraki** [reply]
Actually i think you just need to forward to the clients the nmea 183 sentences the maina pcb sends. With asynchronous web server so many clients can connect. I could make the esp programm if someone can take care of the website. I cant write html.

**Reply by muebau** [reply]
This is all already done in our project. The firmware works more or less perfect. We thought to cleanup the code, write a better documentation and release it then but never found the time to do so.

There were minor things to implement but all main functions worked.

**Reply by kagouraki** [reply]
How is the pinout ? I want to try it.

**Reply by muebau** [reply]
Well if you use a ESP32 D1 mini (clone) you just need to connect the AIS on "Serial2" (IO17 TX, IO16 RX, if I recall it correctly) and use the USB port to power the device. This way you will get the AIS data too (via USB).

**Reply by peterantypas** [reply]
One of the things I built into my custom Raspberry Pi NMEA router was an AIS cache. It maintains the entire AIS "map" in memory by caching the latest message by MMSI and message type. Then, when a TCP client connects, it dumps the whole map instantly before continuing with incremental updates.

The main benefit of this is that I don't have to keep my iPad on all the time and drain the battery. I can simply push the button to lock it, then unlock and get an instant update.

I will try porting this into the ESP32 to see what happens. There's not nearly as much RAM available, though, so I may need to change the data representation to raw packets instead of NMEA sentences.

**Reply by peterantypas** [reply]
Also, having the map in memory enables in-device collision prediction (which I did experiment with back in the day). Just something to keep in mind ...

**Reply by muebau** [reply]
This sounds very great. I thought about cool things to implement (GPS fix, time, etc.). I like this idea a lot :smile:

**Reply by LundSoftwares** [reply]
Any updates to your code @muebau ?
I managed to hardcode some settings and some other small tweaks and now got a fully working Wifi (Running as STA) module to my MAIANA.
The part that's not fully working (for me anyway) is the ConfigMode, often the Webserver aint started. And when the Webserver is started I cant read out the settings from the Module itself, the info is just from the "Void testParsing".
But the AIS module data might not be available through the regular UART? Do I need to connect the MAIANA BreakoutBoard USB to the ESP32 aswell?

I may take a look myself but if you already have som fixes it seems like a waste of time :)
My main goal was to get the AIS data out to SignalK and then to N2K and that's working as a charm!
Thanks a lot!

![image](https://github.com/peterantypas/maiana/assets/23386303/d5bd6940-ae42-42db-ad3a-3393ec1354b6)

**Reply by muebau** [reply]
As we talked directly I would like to update the info here anyway. I started the project for a friend which lost interest in it. Due to other projects the code did not get any attention.

After I have seen the approach of @peterantypas I think my code is just good for ideas 😉 than for production. I still love the idea of an official MAIANA WiFi adapter.
Thinks like the cache like mentioned by Peter would even add more value to the system.

**Comment by jafrado** [comment]
Sweet! I jumped on the ESP bandwagon this year, ESP8266, ESP32-WROOM and
recently the ESP32-S3.
I've done 2 4 layer board designs with these already  - ESP32-S3 and
platformio are very powerful tools
for this and a great starting point. One learning - ESP32-WROOM is not
recommended for new designs.
And among the ESP, stay away from the PCB antenna if you want any range (I
moved to WROOM for
u.FL versus PCB antenna on 8266).

The ESP32-S3-WROOM-1U-N8RCT-ND will be the next module I use in any design
- $3.98 on digikey and
you can smt that module with a heat gun and attach 802.11 via u.FL. use a
good regulator and a 100uF
cap on the power rail.

Thank you Peter for the great project!

**Comment by peterantypas** [comment]
I finally had some time to try a couple of ideas. Here is an antenna I found. It's tuned for US railroad VHF frequencies, which is very close to AIS for inland use. The antenna is a 1/4 wave so it needs a ground plane:

![IMG_1771](https://user-images.githubusercontent.com/1565933/209481162-2001bde8-aacd-4307-b36e-4443e0fdc310.JPG)

(That PCB was meant for a Swarm satellite modem, but it's a good placeholder for this test)

With this amount of ground plane, the tuning is more or less OK:

![Screen Shot 2022-12-25 at 12 30 38 PM](https://user-images.githubusercontent.com/1565933/209481232-1bff16d5-4e89-484d-b54a-59c61a62fb21.png)

This opens up an interesting possibility. I can create a small form factor, wireless, battery powered MAIANA with a detachable antenna. I don't expect it to perform as well as her big sister but it might be good enough for things like day fishing or boat chartering. I'm going to play with this more over the holiday break.

**Reply by maianawifi** [reply]
@peterantypas 
Very interesting. Would be great if the Maiana Wifi with Option 2 will be available in the future.  It is exactly the thing one needs
for boat chartering getting a chart plotter not providing Wifi.
Guess that navigation apps like Navionics needs NMEA 0183 sentences (instead of raw packets).

**Comment by LundSoftwares** [comment]
@peterantypas
If I understood correctly, the code on top of this post is running on a ESP32 based board.
If that is correct, have  you published (or plan to publish) the code somewhere for test?
I saw your comment regarding a WiFi branch, but that did not include some ESP32 code, right? At least I could not find any. ino or platformIO kind of code 😊
